Remotion/Validation/UnitTests/DefaultValidatorExtensionsTest.cs

```diff
@@ -1,14 +1,448 @@
+      addingPropertyValidationRuleBuilderStub
+          .Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
+          .Callback<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>(func => createdValidator = func.Invoke(initParameters));
+
+      addingPropertyValidationRuleBuilderStub.Object.NotEmpty();
+
+      Assert.That(createdValidator, Is.InstanceOf<NotEmptyStringValidator>());
+      var notEmptyValidator = (NotEmptyStringValidator)createdValidator;
+      Assert.That(notEmptyValidator.ValidationMessage, Is.SameAs(validationMessage));
+    }
+
+    [Test]
+    public void NotEmpty_ForCollection_ReturnsNotEmptyCollectionValidator ()
+    {
+      var validationMessage = new InvariantValidationMessage("Fake message");
+      var initParameters = new PropertyValidationRuleInitializationParameters(validationMessage);
+      IPropertyValidator createdValidator = null;
+      var addingPropertyValidationRuleBuilderStub = new Mock<IAddingPropertyValidationRuleBuilder<object, ICollection>>();
+      addingPropertyValidationRuleBuilderStub
+          .Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
+          .Callback<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>(func => createdValidator = func.Invoke(initParameters));
+
+      addingPropertyValidationRuleBuilderStub.Object.NotEmpty();
+
+      Assert.That(createdValidator, Is.InstanceOf<NotEmptyCollectionValidator>());
+      var notEmptyValidator = (NotEmptyCollectionValidator)createdValidator;
+      Assert.That(notEmptyValidator.ValidationMessage, Is.SameAs(validationMessage));
+    }
+
+    [Test]
+    public void NotEmpty_ForCollection_ReturnsNotEmptyGenericCollectionValidator ()
+    {
+      var validationMessage = new InvariantValidationMessage("Fake message");
+      var initParameters = new PropertyValidationRuleInitializationParameters(validationMessage);
+      IPropertyValidator createdValidator = null;
+      var addingPropertyValidationRuleBuilderStub = new Mock<IAddingPropertyValidationRuleBuilder<object, ICollection<object>>>();
+      addingPropertyValidationRuleBuilderStub
+          .Setup(_ => _.SetValidator(It.IsAny<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>()))
+          .Callback<Func<PropertyValidationRuleInitializationParameters, IPropertyValidator>>(func => createdValidator = func.Invoke(initParameters));
+
+      addingPropertyValidationRuleBuilderStub.Object.NotEmpty();
+
+      Assert.That(createdValidator, Is.InstanceOf<NotEmptyCollectionValidator>());
+      var notEmptyValidator = (NotEmptyCollectionValidator)createdValidator;
+      Assert.That(notEmptyValidator.ValidationMessage, Is.SameAs(validationMessage));
+    }
+
+    [Test]
+    public void NotEmpty_ForCollection_ReturnsNotEmptyReadOnlyCollectionValidator ()
```

Comment:
As discussed, we also need to add tests for `List<T>` to show that types such as ObjectList<T> which implement  `IList`, `IList<T>`, and `IReadOnlyList<T>` work in the overload resolution.